src/System.Memory/src/System/Buffers/ArrayMemoryPool.ArrayMemoryPoolBuffer.cs

```diff
@@ -72,11 +72,22 @@ public sealed override MemoryHandle Pin(int byteOffset = 0)
                 {
                     Retain(); // this checks IsDisposed
 
-                    if (byteOffset != 0 && (((uint)byteOffset) - 1) / Unsafe.SizeOf<T>() >= _array.Length)
-                        ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.byteOffset);
+                    try
+                    {
+                        if ((IntPtr.Size == 4 && (uint)byteOffset > (uint)_array.Length * (uint)Unsafe.SizeOf<T>())
```

Comment:
I didn't use the internal `NUint` struct because otherwise it would require messing with the test project. We can use that type if truly desired, but it seemed like more work than having these two lines.